# Introduction to Python Pre-Processing

## By John Chang [GitHub](https://eve-ning.github.io/)

## Libraries

Libraries are like your trusted brands. They provide you with the best of the best tools (for free)

We'll be using
- **pandas**
- **numpy**

*Rule of Thumb: pandas for 2D things (like Excel), numpy for nD things (for more general cases)*

A specific package for your project is better than a package that deals with a wide array of things!

## Dataset (telecom_churn.csv)

The data we'll be using will be bundled together with these files, feel free to look at it in notepad

## Translating Thoughts Into Programming

In my years of programming, I always turn to my trusty tool, that is English. I'll show you how to program by googling

## Viewing the Data

The very first step would be to look at the data.

Google: python load csv pandas

In [1]:
import pandas

df = pandas.read_csv("telecom_churn.csv")

As **convention**, users usually indicate loaded pandas **DataFrame** as df, or anything similar. This is up to you though!

There's also a way to view the data

You can *technically* print it but...

In [2]:
# print(df)

That looks horrible! Can we do better?

What you want is to get a taste of the data,

thus, you get the first few rows, the **head** of the data

In [3]:
df.head()

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False

## Selecting

Let's say we want to isolate, remove some columns/rows

We can do those here!

### Column

In [4]:
df['State'].head()

0    KS
1    OH
2    NJ
3    OH
4    OK
Name: State, dtype: object

### Row

In [5]:
df.iloc[1].head()

State                  OH
Account length        107
Area code             415
International plan     No
Voice mail plan       Yes
Name: 1, dtype: object

## Tutorial 3.1

Chris from Dep A: ```Get the states and area codes of the file, we also want to know if the customer has churned or not. Can you also export it as out.csv and send it to me?```


**Tip: No rush, use multiple lines to debug and ensure your code is working as intended!**

\begin{align}
Double\, Click\, To\, Reveal\, Tip\, (1)
%Seems like you need to select multiple columns, Google pandas select multiple columns
\end{align}

\begin{align}
Double\, Click\, To\, Reveal\, Tip\, (2)
%You need to export it as a csv, google pandas to csv
\end{align}

\begin{align}
Double\, Click\, To\, Reveal\, Answer
%df_filter = df[['State', 'Area code', 'Churn']]
%df_filter.to_csv("out.csv", index=False); # index False prevents python from printing the row numbers
%df_test = pandas.read_csv("out.csv")
%df_test.head()
\end{align}

In [6]:
# Your Code here

In [7]:
# Test if you exported it correctly!
# df_test = pandas.read_csv("out.csv")
# df_test.head()

## Splitting

Splitting is quite common if you need to split a string.

The current DataFrame doesn't require splitting, so I'll create one that requires splitting

## Tutorial 3.2

This is a hard tutorial because you start from nothing.

This aims to test your Googling, don't worry, it should be a few clicks away!

Chris from Talent Acquisition : `Hey I have this suuuuuuuper long text file but everything is split by an underscore, can you help me with this python script?`

`Here's a small excerpt, it's 10,000,000 entries long actually, it's called df_a`

`Don't worry about the rest of the script`

\begin{align}
Double\, Click\, To\, Reveal\, Tip\, (1)
%You can google this too!
%Here's an example: dataframe["column_to_split"].str.split("_", n = 1, expand = True)
\end{align}

\begin{align}
Double\, Click\, To\, Reveal\, Answer
%df_a["NAME_AGE"].str.split("_", n = 1, expand = True)
\end{align}

In [8]:
df_a = \
    pandas.DataFrame({
        'NAME_AGE':['JIN_33', 'ALBERT_51', 'KURT_13', 'ZEWEN_9']
    })

In [9]:
# Chris' script

### GLOBAL CONSTANTS ###

USER_HEAD = ['Bonnie DA', 'Kaht DT', 'Jeremy LU']
PARTICIPANTS = 354
PARTICIPANTS_MAX = 599

### < BEGIN > READ df_a from SQL DATABASE ###
### < REDACTED >
### < END >   READ df_a from SQL DATABASE ###

# Help me here!
df_a 

# df_a['NAME'].head()
# PLACEHOLDER
### TODO: Script to e-mail Joseph

NAME_AGE
0     JIN_33
1  ALBERT_51
2    KURT_13
3    ZEWEN_9

## Aggregating

Aggregating takes some time to wrap your head around, but no worries, we'll learn through examples.

Let's recap on telecom_churn.csv

Do you remember how to get the first few rows?

In [10]:
# df.<function>

\begin{align}
Double\, Click\, To\, Reveal\, Answer
%df.head()
\end{align}

Let's start by simplifying the dataframe, I'll grab a small chunk of it

In [11]:
df_f = df.iloc[:,0:4] # This is a simpler way of getting the first 4 columns!
df_f.head(15)         # Specifying a number gives you that number of rows instead of 5 (default)

State  Account length  Area code International plan
0     KS             128        415                 No
1     OH             107        415                 No
2     NJ             137        415                 No
3     OH              84        408                Yes
4     OK              75        415                Yes
5     AL             118        510                Yes
6     MA             121        510                 No
7     MO             147        415                Yes
8     LA             117        408                 No
9     WV             141        415                Yes
10    IN              65        415                 No
11    RI              74        415                 No
12    IA             168        408                 No
13    MT              95        510                 No
14    IA              62        415                 No

Here are some common questions:

*\"What's the average Account Length for each state?"*

*\"What's the most common Area code?"*

Or in a more roundabout manner:

*\"Which state should we invest in if we want to retain the user the longest?"*

If the question requires you to somehow **group** similar values, you think of `groupby`

## Example 3.1 - 3.4

Just going to throw some examples at you so you get a gist

In [12]:
df_f.groupby('State').mean().head()

Account length   Area code
State                            
AK          97.192308  438.692308
AL          98.025000  430.625000
AR          96.981818  439.254545
AZ         100.468750  432.656250
CA          99.235294  441.500000

In [13]:
df_f.groupby('International plan').count()

State  Account length  Area code
International plan                                  
No                   3010            3010       3010
Yes                   323             323        323

You can group by 2 things

In [14]:
df_f.groupby(['State', 'International plan'])['Account length'].mean().head(10)

State  International plan
AK     No                     94.666667
       Yes                   127.500000
AL     No                     98.819444
       Yes                    90.875000
AR     No                     95.000000
       Yes                   108.625000
AZ     No                     99.754098
       Yes                   115.000000
CA     No                     96.700000
       Yes                   118.250000
Name: Account length, dtype: float64

Always remember you can **chain** commands

In [15]:
df_f.groupby('State').max().head(3)

Account length  Area code International plan
State                                              
AK                177        510                Yes
AL                200        510                Yes
AR                185        510                Yes

In [16]:
df_f.groupby('State').max()['Account length'].head(3) # This selects Account length column only

State
AK    177
AL    200
AR    185
Name: Account length, dtype: int64

### Quick Tip

Sometimes jupyter prints things as text instead of the pretty table as shown by the previous example!

This is because it's not detected as a **DataFrame**

Let's look at the above example

In [17]:
type(df_f.groupby('State').max()['Account length'].head(3))

pandas.core.series.Series

To fix this, you just need to wrap it with `DataFrame`

In [18]:
pandas.DataFrame(df_f.groupby('State').max()['Account length'].head(3))

Account length
State                
AK                177
AL                200
AR                185

## Tutorial 3.3

James: `Hey can you find out which State has the least Customer Service Calls based on the telecom_churn.csv I sent you? I want to borrow some employees for this weekend for another project.`

**Hint: You should use `df` and filter useful columns**

\begin{align}
Double\, Click\, To\, Reveal\, Tip\, (1)
%Remember how to select columns? You need to select the State and Customer service calls columns
\end{align}

\begin{align}
Double\, Click\, To\, Reveal\, Tip\, (2)
%What methods can you use to find out the least? Note that min isn't the only method feasible!
\end{align}

\begin{align}
Double\, Click\, To\, Reveal\, Answer
%df[['State', 'Customer service calls']].groupby('State').mean().sort_values('Customer service calls')
%You can use min or median, but is it as useful as mean
\end{align}

## Unique

This is self-explanatory!

Unique basically gets unique entries, also known as distinct entries

I'll create another DataFrame

In [19]:
df_g = df[['State', 'Area code', 'Churn']]

In [20]:
df_g.head(15)

State  Area code  Churn
0     KS        415  False
1     OH        415  False
2     NJ        415  False
3     OH        408  False
4     OK        415  False
5     AL        510  False
6     MA        510  False
7     MO        415  False
8     LA        408  False
9     WV        415  False
10    IN        415   True
11    RI        415  False
12    IA        408  False
13    MT        510  False
14    IA        415  False

How many **distinct** area codes are there?

A good guess is that there are 3, [408, 415, 510].

We can find out by calling `unique` on the column though

In [21]:
df_g['Area code'].unique()

array([415, 408, 510], dtype=int64)

## Tutorial 3.4

Quick tutorial, get all unique States!

\begin{align}
Double\, Click\, To\, Reveal\, Answer
%df_g['State'].unique()
\end{align}

## Joining

Sometimes you'd get 2 tables that are related somehow, here's an example


## Example 3.5
Let's say you're tasked to get the total cost of order `df_h`

Your database `df_i` holds all of the price data.

Intuitively, you need to join/merge/combine them!

*The order could've been imported in from a csv!*

In [22]:
df_h = pandas.DataFrame({
    'Count' : [1, 32, 18],
    'ItemID' : [1, 3, 4]
})
df_h

Count  ItemID
0      1       1
1     32       3
2     18       4

In [23]:
df_i = pandas.DataFrame({
    'ItemID' : [1, 2, 3],
    'ItemName' : ['Choco', 'Lime', 'Strawberry'],
    'Cost' : [34.4, 14.5, 62.4]
})
df_i

ItemID    ItemName  Cost
0       1       Choco  34.4
1       2        Lime  14.5
2       3  Strawberry  62.4

It's as simple as `merge()`

The function finds all similar column names and joins them!

In [24]:
pandas.merge(df_i, df_h)

ItemID    ItemName  Cost  Count
0       1       Choco  34.4      1
1       3  Strawberry  62.4     32

Notice that some entries are missing!

If the database was a few thousand entries long, it'll be hard to trace where the entries have disappeared to

### Different Merges

For most cases, this may suffice as is, but sometimes you need a different merge, let's explore other options

`merge(LEFT, RIGHT, METHOD)`

#### Left Join

This means **I want to keep all of the left entries**. (The left is the 1st argument in the `merge()` function)

If the **right entry** doesn't exist, we put a dummy value

In [25]:
pandas.merge(df_i, df_h, how='left') 

ItemID    ItemName  Cost  Count
0       1       Choco  34.4    1.0
1       2        Lime  14.5    NaN
2       3  Strawberry  62.4   32.0

#### Right Join

This means **I want to keep all of the right entries**. (The right is the 2nd argument in the `merge()` function)

If the **left entry** doesn't exist, we put a dummy value

In [26]:
pandas.merge(df_i, df_h, how='right')

ItemID    ItemName  Cost  Count
0       1       Choco  34.4      1
1       3  Strawberry  62.4     32
2       4         NaN   NaN     18

#### Outer Join

It's a combo of both Left and Right Join. Outer means I want **all** entries regardless of if it matches

In [27]:
pandas.merge(df_i, df_h, how='outer')

ItemID    ItemName  Cost  Count
0       1       Choco  34.4    1.0
1       2        Lime  14.5    NaN
2       3  Strawberry  62.4   32.0
3       4         NaN   NaN   18.0

#### Inner Join

This is the default option. The function drops/removes any rows that doesn't have a match!

In [28]:
pandas.merge(df_i, df_h, how='inner') # == pandas.merge(df_i, df_h)

ItemID    ItemName  Cost  Count
0       1       Choco  34.4      1
1       3  Strawberry  62.4     32

## Tutorial 3.5

Matthew: `here's the csv June KT prepared (df_j), compare with out company's deliverable Database (df_k) and report the total CONFIRMED construction cost back pls`

*There are intentional problems in the DataFrames!*

\begin{align}
Double\, Click\, To\, Reveal\, Tip\, (1)
%If you've tried merging, you'd find yourself a long error, read the column headers carefully and debug!
\end{align}

\begin{align}
Double\, Click\, To\, Reveal\, Tip\, (2)
%You might be stuck on how to multiply count with the cost. What are the right terms to google?
%'Mutliply 2 columns together' is a good start!
\end{align}

\begin{align}
Double\, Click\, To\, Reveal\, Tip\, (3)
%How do you sum a column? It's as your logic would carry you, grab the column, then you sum it!
\end{align}

\begin{align}
Double\, Click\, To\, Reveal\, Answer
%df_order_company = pandas.merge(df_order, df_company, left_on='DelivrableID', right_on='DeliverableID')
%df_order_company['CountCost'] = df_order_company['Count'] * df_order_company['Cost']
%df_order_company = df_order_company[df_order_company['Confirmed']]
%df_order_company['CountCost'].sum()
\end{align}

In [29]:
df_order = pandas.DataFrame({
    'Count'         : [1, 26, 1, 18, 4],
    'DelivrableID'  : [1, 3,  4, 6, 10],
    'Confirmed'     : [True, True, True, False, True]
})

In [30]:
df_company = pandas.DataFrame({
    'DeliverableID'   : [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
    'DeliverableName' : ['Project Alpha Planning',
                         'Project Beta Planning',
                         'Workers/Contractors',
                         'Construction Delay/Month',
                         'Construction Ext./Month',
                         'Food/Drinks',
                         'Extra PPE',
                         'External Serv. (Toilet)',
                         'External Serv. (Catering)',
                         'External Serv. (Transport)'],
    'Cost'            : [3300, # 'Project Alpha Planning',
                         3560, # 'Project Beta Planning',
                         1450, # 'Workers/Contractors',
                         1900, # 'Construction Delay/Month',
                         1750, # 'Construction Ext./Month',
                         100 , # 'Food/Drinks',
                         230 , # 'Extra PPE',
                         200 , # 'External Serv. (Toilet)',
                         340 , # 'External Serv. (Catering)',
                         100 ] # 'External Serv. (Transport)'
})

# Other Functions

There are some other functions I find useful to learn

```
Sorting
Slicing (Selecting Multiple Columns/Rows Instantly)
Transposing
Filtering by Values
Handling Missing Values
```

Learning Python itself helps a lot too, don't restrict yourself to Data Science stuff only!

# End

This is the end of this part of the course!

Next up is Visualizing Data